pkg/storage/unified/sql/db/service.go

```diff
@@ -59,3 +63,69 @@ type ContextExecer interface {
 	QueryContext(ctx context.Context, query string, args ...any) (*sql.Rows, error)
 	QueryRowContext(ctx context.Context, query string, args ...any) *sql.Row
 }
+
+// WithTxFunc is an adapter to be able to provide the DB.WithTx method as an
+// embedded function.
+type WithTxFunc func(context.Context, *sql.TxOptions, TxFunc) error
+
+// WithTx implements the DB.WithTx method.
+func (x WithTxFunc) WithTx(ctx context.Context, opts *sql.TxOptions, f TxFunc) error {
+	return x(ctx, opts, f)
+}
+
+// BeginTxFunc is the signature of the DB.BeginTx method.
+type BeginTxFunc = func(context.Context, *sql.TxOptions) (Tx, error)
+
+// NewWithTxFunc provides implementations of DB an easy way to provide the
+// DB.WithTx method.
+// Example usage:
+//
+//	type myDB struct {
+//		db.WithTxFunc // embedded so that `WithTx` is already provided
+//		// other members...
+//	}
+//
+//	func NewMyDB(/* options */) (db.DB, error) {
+//		ret := new(myDB)
+//		ret.WithTxFunc = db.NewWithTxFunc(ret.BeginTx)
+//		// other initialization code ...
+//		return ret, nil
+//	}
+func NewWithTxFunc(x BeginTxFunc) WithTxFunc {
```

Comment:
Moved here from `dbimpl/db.go` since this can be reused by other implementations of `DB`, and we do reuse it in the `otel` package.